### PR TITLE
First draft of an Offline Esimator

### PR DESCRIPTION
**Pull request by WardLT**
This PR adds a first attempt at an offline estimation strategy: computing the mean-squared error loss then optimizing parameters using a SciPy optimizer.

It adds:
- A base class for the estimator modeled after the `OnlineEstimator`
- A similar pattern for building wrappers around loss functions used in offline estimation to those used for online
- An initial documentation page
- Use of offline in the CAMP notebook

Todo:
- [x] API documentation for offline estimation
- [x] Consulting with Noah to see if the pattern works with his style
- [x] *Adding detailed docs for Objectives and Offline Estimator types* (waiting until I have >1)

### CODE REVIEW

**Review by victorventuri — COMMENTED** [review]
It's not easy to write a comment on the notebook in the diff, so I will write it here:
I like the use of the ScipyMinimizer to better initialize the UKF, but I find it a bit weird that, after the fit, it still shows poor performance. I know we are trying to not-overfit, but wouldn't it be a more representative example to fit well to the RPT section, and use those for initializing the UKF?

Overall, I think these are great! I would just prefer if we altered some terminology to align with that from the `OnlineEstimator`

File: moirae/estimators/offline/objectives.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import numpy as np
+from batdata.data import BatteryDataset
+
+from moirae.interface import row_to_inputs
+from moirae.models.base import CellModel, HealthVariable, GeneralContainer
+from moirae.simulator import Simulator
+
+
+# TODO (wardlt): Add degradation model after we decide how to handle its parameters
+class Objective:
```
[the remark this answers is not in the document]
Comment (WardLT) [reply moirae/estimators/offline/objectives.py:12]: Loss is a good idea. It's clear that lower is better

File: moirae/estimators/offline/objectives.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    """Observed data for the battery performance"""
+
+    def __init__(self,
+                 cell_model: CellModel,
+                 asoh: HealthVariable,
+                 state: GeneralContainer,
+                 observations: BatteryDataset):
+        self.cell_model = cell_model
+        self.asoh = asoh.model_copy(deep=True)
+        self.state = state.model_copy(deep=True)
```
[the remark this answers is not in the document]
Comment (WardLT) [reply moirae/estimators/offline/objectives.py:34]: Yea, adding "transient" should cut down on a lot of confusion

File: moirae/estimators/offline/scipy.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    Args:
+        objective: Objective function to be optimized
+        kwargs: Passed to the minimize function
+    """
+
+    def __init__(self,
+                 objective: Objective,
+                 **kwargs):
+        self.objective = objective
+        self.scipy_kwargs = kwargs
```
[the remark this answers is not in the document]
Comment (WardLT) [reply moirae/estimators/offline/scipy.py:22]: Good point. I'll provide an code example which shows changing a few of them and emphasize that the page linked in the class docstring explains the keyword arguments. 

The TypedDict is a cool idea, but I'm leery about keeping it up-to-date with SciPy

File: moirae/estimators/offline/scipy.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    def estimate(self) -> tuple[GeneralContainer, HealthVariable, OptimizeResult]:
+        # Get the scale of the initial error, used to normalize the output
+        x0 = self.objective.get_x0()
+        y0 = self.objective(x0[None, :]).item()  # Should be a scalar
+
+        # Assemble the function call
+        result = minimize(
+            fun=lambda x: self.objective(x[None, :]).item() / max(y0, 1e-12),
```
[the remark this answers is not in the document]
Comment (WardLT) [reply moirae/estimators/offline/scipy.py:31]: Numerical reasons. I want the values passed to the optimizer to be close to 1, so I make it such that the initial guess is 1. I'm also not 100% on whether all SciPy optimizers use relative or absolute tolerance for their cutoff

[the remark this answers is not in the document]
Comment (WardLT) [reply moirae/estimators/offline/scipy.py:31]: Seeing your next comment, it seems a bad idea to assume that y is positive. Maybe I need to rethink this or `abs(y0)`

File: moirae/estimators/offline/objectives.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        """
+        Compute the fitness of a set of parameters to the observed data.
+
+        Args:
+            x: Batch of parameters to be evaluated (2D array)
+
+        Returns:
+            A single measure of parameter fitness for each member of batch (1D array)
+        """
+        raise NotImplementedError()
```
[the remark this answers is not in the document]
Comment (WardLT) [reply moirae/estimators/offline/objectives.py:57]: Great point. I suppose I do assume that it's positive in the SciPy optimizer. Maybe I need to remove that assumption

File: tests/conftest.py [hunk 6]
```diff
@@ -46,7 +46,7 @@ def timeseries_dataset(simple_rint) -> BatteryDataset:
 
     def _update_outputs(time, current, transient, outputs):
         output['test_time'].append(time)
-        output['current'].append(current)
+        output['current'].append(-current)  # Battery data toolkit uses opposite sign convention
```
[the remark this answers is not in the document]
Comment (WardLT) [reply tests/conftest.py:49]: This frightens me too. I think I only used the "timeseries_dataset" one place, and I never checked if the output of the state estimation was reasonable

File: moirae/estimators/offline/scipy.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (victorventuri) [reply moirae/estimators/offline/scipy.py:31]: Yeah, I figured it was to avoid some numerical difficulties. Would this cause problems if the initial guess is very poor? (for example, such that the computer-precision division by `y0` makes two different cost values appear the same to the optimizer)

File: moirae/estimators/offline/scipy.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (victorventuri) [reply moirae/estimators/offline/scipy.py:22]: I had not considered that aspect... at the same time, if we don't want to deal with that, can we not just add the current version of scipy in the `.toml` as a dependency?

[the remark this answers is not in the document]
Comment (victorventuri) [reply moirae/estimators/offline/scipy.py:22]: (I say that because, under the same principle, we don't know that `scipy` won't change its structure such that the `minimize` function is actually under a different submodule, for example, even though I know that is extremely unlikely to ever happen haha)

File: tests/conftest.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (victorventuri) [reply tests/conftest.py:49]: I thought I had at some point either corrected that, or fixed the test somewhere... so idk, still puzzled by this one; we might need to check the history of merges and commits to see what happened

[the remark this answers is not in the document]
Comment (WardLT) [reply tests/conftest.py:49]: The [blame on main](https://github.com/ROVI-org/auto-soh/blame/main/tests/conftest.py) doesn't show that it has been changed to/from the sign convention. It isn't my only time messing up sign changes, so 🤷

File: moirae/estimators/offline/scipy.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (WardLT) [reply moirae/estimators/offline/scipy.py]: Great question. I'm not sure. Shall I make it easy to turn "loss scaling" on and off?

File: moirae/estimators/offline/scipy.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (WardLT) [reply moirae/estimators/offline/scipy.py:23]: I'm more against pinning the scipy version than writing the TypedDict, but writing the typeddict just feels like a lot of tedious work which (there are tons of optimizer options)

[the remark this answers is not in the document]
Comment (victorventuri) [reply moirae/estimators/offline/scipy.py:23]: yes, absolutely, `scipy` is very well documented and I am just being a bit of a cry-baby haha this is perfectly fine 😄

File: moirae/estimators/offline/scipy.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (victorventuri) [reply moirae/estimators/offline/scipy.py]: I think that's a great idea!

File: tests/conftest.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (victorventuri) [reply tests/conftest.py:49]: (though that should be a separate PR imo)
